### ISSUE TITLE
allowing multiple active scripts

### ISSUE THREAD

**Issue by HuStmpHrrr** [ProofGeneral/PG#199]
I am wondering is there any specific reason why there must be at most one single active script at any instant? 

quote http://proofgeneral.inf.ed.ac.uk/htmlshow.php?title=Proof+General+user+manual&file=releases%2FProofGeneral%2Fdoc%2FProofGeneral%2FProofGeneral_4.html#Switching-between-proof-scripts
>However, you cannot have more than one buffer where only a fraction of the proof script contains a locked region. Before you can employ script management in another proof script buffer, you must either fully assert or retract the current script buffer.

quote https://proofgeneral.github.io/doc/userman/ProofGeneral_3/#Active-scripting-buffer
>If you have a partly processed scripting buffer and use C-c C-s, or you attempt to use script processing in a new buffer, Proof General will ask you if you want to retract what has been proved so far, Scripting incomplete in buffer myproof.l, retract? or if you want to process the remainder of the active buffer, Completely process buffer myproof.l instead? before you can start scripting in a new buffer. If you refuse to do either, Proof General will give an error message: Cannot have more than one active scripting buffer!. 

I don't understand why it's designed to work in this way. coqide doesn't seem to have this problem. Is this some sort of implementation limitation? I'd imagine it's not normal for one to run multiple scripts at the same time.

**Comment by cpitclaudel** [comment]
> I'd imagine it's not normal for one to run multiple scripts at the same time.

I think it's pretty normal, actually.  I'd guess it's mostly historical reasons. I don't know how hard it would be to change things.

There's also a desire to keep things consistent. If you compile and load a dependency of your current file, then that module will be marked as fully processed and locked.  That's useful from preventing dependencies from being modified without retracting the things that depend on them first.

**Comment by HuStmpHrrr** [comment]
@cpitclaudel true. though considering the power of emacs, and how people use it, say, running emacs demo, having to keep a single active script is a too strong limitation. the scripts might not be related, or otherwise the user might just know what he's trying to do.

**Comment by Matafou** [comment]
We could try to have several coq session in parallel (this would need a deep refactoring but it is not impossible) but how would we know they are independent? Suppose you have to sessions in parallel on file1.v and file2.v. Suppose you type in the middle of file2.v "Require Import file1." then what should we do? End the coq session on file1.v and lock it? Let it go and have the one on file2.v be out of sync with file1.v?

I am not sure the feature is worth the effort anyway. Given the power of computers today running two independent emacs for that is not a problem.

**Comment by HuStmpHrrr** [comment]
@Matafou i am not sure what's the problem though. I would assume having two emacs will become an out of sync solution. In fact, I don't think coqide does something special either.

**Comment by hendriktews** [comment]
HuStmpHrrr <notifications@github.com> writes:

> I am wondering is there any specific reason why there must be at most
> one single active script at any instant?

There are a number of reasons that make it much simpler to
support just one active scripting buffer. One that has been
mentioned already is the consistency of the whole development.
Another one is that the code is easier to organize around a
global instance of the Coq background process. It would require
some effort to overcome these limitations, but it would not be
impossible.

I, however, don't see why you would want two active scripting
buffers. Are you really working in parallel in two independent
developments? I often work in several files, for instance, a.v
and b.v, but then one depends on the other, for instance, b
imports a. When I go back to b, then I want to lock a, in order
to see that the stuff that I added to a is useful for what I do
in b. When I go to a, then only because some definition or lemma
there needs adjustment for what I am doing in b. Then I want that
b is retract.

Bye,

Hendrik

**Comment by Matafou** [comment]
We could have two active coq processes in independent files that depend on
the same set of files, is it this situation you have? My experience is that
it is quite uncommon but it happens some times.

If we except this situation either you edit files that are dependent of
each other and PG is ok (especially with autocompilation on). Or you are
editing two independent sets of files and launching two emacs is ok.

P.


2017-09-03 2:26 GMT+02:00 HuStmpHrrr <notifications@github.com>:

**Comment by HuStmpHrrr** [comment]
@Matafou not quite but close. I keep an emacs daemon running and simply attach to the server when I need to. so effectively indeed I edit multiple projects at the same time.

**Comment by ybertot** [comment]
@hendriktews The reason I know to want having several active scripts at a time is for porting reasons: play the old proof and the new one at the same time (practically in lock step) to see where discrepancies in behavior appear.

**Comment by hendriktews** [comment]
Yves Bertot <notifications@github.com> writes:

> play the old proof and the new one at the same time

Good point.
